xwiki-platform-core/xwiki-platform-livedata/xwiki-platform-livedata-webjar/src/main/webjar/filters/livedata-filter.html

```diff
@@ -0,0 +1,137 @@
+<template>
+  <div class="livedata-filter">
+    <component
+      v-if="componentName"
+      :is="componentName"
+      :property-id="propertyId"
+      :index="index"
+      :logic="logic"
+    ></component>
+  </div>
+</template>
+
+
+<script>
+/*
+  * See the NOTICE file distributed with this work for additional
+  * information regarding copyright ownership.
+  *
+  * This is free software; you can redistribute it and/or modify it
+  * under the terms of the GNU Lesser General Public License as
+  * published by the Free Software Foundation; either version 2.1 of
+  * the License, or (at your option) any later version.
+  *
+  * This software is distributed in the hope that it will be useful,
+  * but WITHOUT ANY WARRANTY; without even the implied warranty of
+  * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+  * Lesser General Public License for more details.
+  *
+  * You should have received a copy of the GNU Lesser General Public
+  * License along with this software; if not, write to the Free
+  * Software Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA
+  * 02110-1301 USA, or see the FSF site: http://www.fsf.org.
+  */
+define([
+  "Vue",
+], function (
+  Vue
+) {
+
+  Vue.component("livedata-filter", {
+
+    name: "livedata-filter",
+
+    template: template,
+
+    props: {
+      propertyId: String,
+      index: Number,
+      logic: Object,
+    },
+
+    data: function () {
+      return {
+        componentName: undefined,
+      };
+    },
+
+    computed: {
+      filterId: function () {
+        return this.logic.getFilterDescriptor(this.propertyId).id;
+      },
+    },
+
+    methods: {
+      loadFilter: function (filterId) {
+        var self = this;
+        return new Promise (function (resolve, reject) {
+
+          var componentName = "filter-" + filterId;
+
+          // load success callback
+          var loadFilterSuccess = function () {
+            self.componentName = componentName;
+            resolve();
+          };
+
+          // load error callback
+          var loadFilterFailure = function (err) {
+            console.warn(err);
+            reject();
+          };
+
+          // load filter based on it's id
+          require(["vue!" + BASE_PATH + "filters/" + componentName + ".html"],
+            loadFilterSuccess,
+            loadFilterFailure
+          );
+
+        });
+      },
+    },
+
+    mounted: function () {
+      var self = this;
+      // load filter
+      this.loadFilter(this.filterId).catch(function () {
+        self.loadFilter("default").catch (function (err) {
```

Comment:
I'm doing the check if the property is filterable before, outside the element as a `v-if` on the tag:
```html
<livedata-filter
  v-if="logic.isPropertyFilterable(col.id)"
  :property-id="col.id"
  ...
></livedata-filter>
```
but indeed I can verify if the property is filterable in this loadFilter function too, for safer code.

Also, we can verify if the property is filterable in the `getFilterDescriptor` method, and return `null` or `{}` if it is not. (I think `{}` would be better as the method would always return the same type, and we already have a method to know if it's filterable or not)

Reply:
> (I think {} would be better as the method would always return the same type ...)

Well, actually that does not make any difference because the method is already return `undefined` for invalid property names.

Reply:
> we can verify if the property is filterable in the `getFilterDescriptor` method, and return null or {} if it is not.

Well actually I can't do that because `getFilterDescriptor` is used by the `isPropertyFilterable` method, that looks if the id is null of not.

> It's not clear to me what happens if the property is not filterable, since it appears you always fallback to default filter.

It is not a problem, because the checks are done before getting to that point.
But if a layout doesn't implement the `<livedata-filter>` element correctly, yes it would appears as a default displayer, but then the logic would block any changes from it, so it would only be a visual issue.